GitHub app blog (#2988)
jenkins-infra/jenkins.io: content/blog/2020/03/2020-03-19-github-app-authentication.adoc

---
layout: post
title: "GitHub app authentication released"
tags:
- github
- github-branch-source
author: timja
---

I'm excited to announce support for authenticating as a GitHub app in Jenkins.

It has been released in link:https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.7.0[GitHub Branch Source 2.7.0]

This has been a long awaited feature by many users,

It brings many benefits:

* the rate limit for a GitHub app scales with your organization size, 
whereas a user based token has a limit of 5000 regardless of how many repositories you have,

> * for organization’s that have 2fa enforced - no need to manage 2fa tokens for a 'bot' user

Reviewer: ```suggestion
* for organizations that have 2FA enforced - no need to manage 2FA tokens for a 'bot' user
```

* to improve and tighten security: the Jenkins GitHub app requires a minimum, controlled set of privileges compared to a service user and its personal access token which has a much wider set of privileges

== Getting started

Install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin],
make sure the version is at least 2.7.0.

> Follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[setup guide], it's also linked to from the plugins README on GitHub.

Reviewer: ```suggestion
Follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[setup guide]; it's also linked to from the plugin’s README on GitHub.
```

> Follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[setup guide]; it's also linked to from the plugin’s README on GitHub.

Reviewer 2: ```suggestion
Follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[setup guide]; it's also linked from the plugin’s README on GitHub.
```

Once you've finished setting it up, Jenkins will validate your credential and you should see your new rate limit,

> here's an example on a large org:

Reviewer: ```suggestion
Once you've finished setting it up, Jenkins will validate your credential and you should see your new rate limit.
Here's an example on a large org:
```

image:/images/github-app-rate-limit.png[GitHub app rate limit]

== How do I get an API token in my pipeline?

> You can access the Bearer token for the GitHub API by just loading a 'Username/ Password' credential as usual,

Reviewer: ```suggestion
You can access the Bearer token for the GitHub API by just loading a 'Username/Password' credential as usual,
```

the plugin will handle authenticating with GitHub in the background:

[source, groovy]
----

pipeline {
  agent any

  stages{
    stage('Check run') { 
      steps {

>         withCredentials([usernamePassword(credentialsId: 'githubapp-jenkins', usernameVariable: 'GITHUB_APP', passwordVariable: 'GITHUB_JWT_TOKEN')]) {

Reviewer 2: Easier to read if the line is broken at strategic locations
```suggestion
        withCredentials([usernamePassword(credentialsId: 'githubapp-jenkins',
                                          usernameVariable: 'GITHUB_APP',
                                          passwordVariable: 'GITHUB_JWT_TOKEN')]) {
```

            sh '''

>             curl -H "Content-Type: application/json" -H "Accept: application/vnd.github.antiope-preview+json" -H "authorization: Bearer ${GITHUB_JWT_TOKEN}" -d '{ "name": "check_run", "head_sha": "'${GIT_COMMIT}'", "status": "in_progress", "external_id": "42", "started_at": "2020-03-05T11:14:52Z", "output": { "title": "Check run from Jenkins!", "summary": "This is a check run which has been generated from Jenkins as GitHub App", "text": "...and that is awesome"}}' https://api.github.com/repos/<org>/<repo>/check-runs

Reviewer 2: Easier to read if the line is broken at strategic locations.
```suggestion
            curl -H "Content-Type: application/json" \
                 -H "Accept: application/vnd.github.antiope-preview+json" \
                 -H "authorization: Bearer ${GITHUB_JWT_TOKEN}" \
                 -d '{ "name": "check_run", \
                       "head_sha": "'${GIT_COMMIT}'", \
                       "status": "in_progress", \
                       "external_id": "42", \
                       "started_at": "2020-03-05T11:14:52Z", \
                       "output": { "title": "Check run from Jenkins!", \
                                   "summary": "This is a check run which has been generated from Jenkins as GitHub App", \
                                   "text": "...and that is awesome"}}' https://api.github.com/repos/<org>/<repo>/check-runs
```

>             curl -H "Content-Type: application/json" -H "Accept: application/vnd.github.antiope-preview+json" -H "authorization: Bearer ${GITHUB_JWT_TOKEN}" -d '{ "name": "check_run", "head_sha": "'${GIT_COMMIT}'", "status": "in_progress", "external_id": "42", "started_at": "2020-03-05T11:14:52Z", "output": { "title": "Check run from Jenkins!", "summary": "This is a check run which has been generated from Jenkins as GitHub App", "text": "...and that iss awesome"}}' https://api.github.com/repos/<org>/<repo>/check-runs

Reviewer: ```suggestion
            curl -H "Content-Type: application/json" -H "Accept: application/vnd.github.antiope-preview+json" -H "authorization: Bearer ${GITHUB_JWT_TOKEN}" -d '{ "name": "check_run", "head_sha": "'${GIT_COMMIT}'", "status": "in_progress", "external_id": "42", "started_at": "2020-03-05T11:14:52Z", "output": { "title": "Check run from Jenkins!", "summary": "This is a check run which has been generated from Jenkins as GitHub App", "text": "...and that is awesome"}}' https://api.github.com/repos/<org>/<repo>/check-runs
```

Reviewer 3: The typo “iss” was there on purpose when I wrote that example 😂. Kind of the same typo for the first Docker demo: 
https://youtu.be/wW9CAH9nSLs?t=254

But it’s a good catch anyway!

            '''
        }
      }
    }
  }
}

----

== What's next

There's a proposed Google Summer of Code project: link:https://jenkins.io/projects/gsoc/2020/project-ideas/github-checks/[GitHub Checks API for Jenkins Plugins]. 

This is looking at integrating with the link:https://developer.github.com/v3/checks/[checks api], with a focus on reporting issues found using the link:https://plugins.jenkins.io/warnings-ng/[warnings-ng plugin] directly onto the GitHub pull request, along with test results summary on GitHub.

> If you want to get involved with this join the link:https://gitter.im/jenkinsci/gsoc-sig[gitter channel] and ask how you can help, there's already students sending in their proposals.

Reviewer 2: ```suggestion
If you want to get involved with this, join the link:https://gitter.im/jenkinsci/gsoc-sig[gitter channel] and ask how you can help.
There are already students sending in their proposals.
```

---
layout: post
title: "GitHub app authentication released"
tags:
- github
- github-branch-source
author: timja
---

I'm excited to announce support for authenticating as a GitHub app in Jenkins.

It has been released in link:https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.7.0[GitHub Branch Source 2.7.0]

This has been a long awaited feature by many users,

It brings many benefits:

* the rate limit for a GitHub app scales with your organization size, 
whereas a user based token has a limit of 5000 regardless of how many repositories you have,

* for organizations that have 2FA enforced - no need to manage 2FA tokens for a 'bot' user

* to improve and tighten security: the Jenkins GitHub app requires a minimum, controlled set of privileges compared to a service user and its personal access token which has a much wider set of privileges

> 

Author: ```suggestion

* access to the GitHub link:https://developer.github.com/v3/checks/[checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation
```

---
layout: post
title: "GitHub app authentication released"
tags:
- github
- github-branch-source
author: timja
---

I'm excited to announce support for authenticating as a GitHub app in Jenkins.

It has been released in link:https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.7.0-beta1[GitHub Branch Source 2.7.0-beta1]

This has been a long awaited feature by many users,

It brings many benefits:

* the rate limit for a GitHub app scales with your organization size, 
whereas a user based token has a limit of 5000 regardless of how many repositories you have,

> * for organizations that have 2FA enforced - no need to manage 2FA tokens for a 'bot' user

Reviewer 4: ```suggestion
* ** User-independent authentication** - Each GitHub app has its own user-independent authentication. No more need for 'bot' users or figuring out who should be the owner of 2FA or OAuth tokens.
```

> * to improve and tighten security: the Jenkins GitHub app requires a minimum, controlled set of privileges compared to a service user and its personal access token which has a much wider set of privileges

Reviewer 4: ```suggestion
* **Improved security and tighter permissions** - GitHub Apps offers much finer-grained permissions compared to a service user and its personal access tokens. This lets the Jenkins GitHub app require a much smaller set of privileges to run properly.

> * access to the GitHub link:https://developer.github.com/v3/checks/[checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation

Author: ```suggestion
* access to the GitHub link:https://developer.github.com/v3/checks/[checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation

```

> * access to the GitHub link:https://developer.github.com/v3/checks/[checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation

Reviewer 4: ```suggestion
* **Access to GitHub Checks API** - GitHub Apps can access the the  link:https://developer.github.com/v3/checks/[GitHub Checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation
```

== Getting started

> Install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin],
> make sure the version is at least 2.7.0-beta1.
> 
> Follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[setup guide]; it's also linked from the plugin’s README on GitHub.

Reviewer 4: ```suggestion
Install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin],
make sure the version is at least 2.7.0-beta1. 
Then follow the link:https://github.com/jenkinsci/github-branch-source-plugin/blob/master/docs/github-app.adoc[setup guide].  These instructions are also linked from the plugin’s README on GitHub.
```

[source, groovy]
----

pipeline {
  agent any

  stages{
    stage('Check run') { 
      steps {
        withCredentials([usernamePassword(credentialsId: 'githubapp-jenkins',
                                          usernameVariable: 'GITHUB_APP',
                                          passwordVariable: 'GITHUB_JWT_TOKEN')]) {
            sh '''
            curl -H "Content-Type: application/json" \
                 -H "Accept: application/vnd.github.antiope-preview+json" \
                 -H "authorization: Bearer ${GITHUB_JWT_TOKEN}" \
                 -d '{ "name": "check_run", \
                       "head_sha": "'${GIT_COMMIT}'", \
                       "status": "in_progress", \
                       "external_id": "42", \
                       "started_at": "2020-03-05T11:14:52Z", \
                       "output": { "title": "Check run from Jenkins!", \
                                   "summary": "This is a check run which has been generated from Jenkins as GitHub App", \
                                   "text": "...and that is awesome"}}' https://api.github.com/repos/<org>/<repo>/check-runs
            '''
        }
      }
    }
  }
}

> ----

Reviewer 4: @timja 
It would be helpful if we talked a bit about why users would want to do this and some of the security considerations in this are. Also, we should definitely mention that this token is only valid for one hour.  For long running jobs if the user gets the password this way, they should use it right away.  This is also a security improvement - tokens visible in the pipeline are not long lived.

== What's next

> There's a proposed Google Summer of Code project: link:https://jenkins.io/projects/gsoc/2020/project-ideas/github-checks/[GitHub Checks API for Jenkins Plugins]. 
> 
> This is looking at integrating with the link:https://developer.github.com/v3/checks/[checks api], with a focus on reporting issues found using the link:https://plugins.jenkins.io/warnings-ng/[warnings-ng plugin] directly onto the GitHub pull request, along with test results summary on GitHub.
> 
> If you want to get involved with this, join the link:https://gitter.im/jenkinsci/gsoc-sig[gitter channel] and ask how you can help.

Reviewer 4: ```suggestion
GitHub Apps authentication in Jenkins is a huge improvement.  Many teams have already started using it and have helped improve it by giving pre-release feedback. There are more improvements on the way. 

There's a proposed Google Summer of Code project: link:https://jenkins.io/projects/gsoc/2020/project-ideas/github-checks/[GitHub Checks API for Jenkins Plugins].  
It will look at integrating with the link:https://developer.github.com/v3/checks/[checks api], 
with a focus on reporting issues found using the link:https://plugins.jenkins.io/warnings-ng/[warnings-ng plugin]
directly onto the GitHub pull requests, along with test results summary on GitHub.
If you want to get involved with this, join the link:https://gitter.im/jenkinsci/gsoc-sig[gitter channel]
and ask how you can help.
```

---
layout: post
title: "GitHub app authentication released"
tags:
- github
- github-branch-source
author: timja
---

I'm excited to announce support for authenticating as a GitHub app in Jenkins.

> It has been released in link:https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.7.0[GitHub Branch Source 2.7.0]

Author: ```suggestion
It has been released in link:https://github.com/jenkinsci/github-branch-source-plugin/releases/tag/github-branch-source-2.7.0-beta1[GitHub Branch Source 2.7.0-beta1]
```

This has been a long awaited feature by many users,

It brings many benefits:

* the rate limit for a GitHub app scales with your organization size, 
whereas a user based token has a limit of 5000 regardless of how many repositories you have,

* for organizations that have 2FA enforced - no need to manage 2FA tokens for a 'bot' user

* to improve and tighten security: the Jenkins GitHub app requires a minimum, controlled set of privileges compared to a service user and its personal access token which has a much wider set of privileges

* access to the GitHub link:https://developer.github.com/v3/checks/[checks API] to create check runs and check suites from Jenkins jobs and provide detailed feedback on commits as well as code annotation

== Getting started

Install the link:https://plugins.jenkins.io/github-branch-source/[GitHub Branch Source plugin],

> make sure the version is at least 2.7.0.

Author: ```suggestion
make sure the version is at least 2.7.0-beta1.
```

                       "head_sha": "'${GIT_COMMIT}'", \
                       "status": "in_progress", \
                       "external_id": "42", \
                       "started_at": "2020-03-05T11:14:52Z", \
                       "output": { "title": "Check run from Jenkins!", \
                                   "summary": "This is a check run which has been generated from Jenkins as GitHub App", \
                                   "text": "...and that is awesome"}}' https://api.github.com/repos/<org>/<repo>/check-runs
            '''
        }
      }
    }
  }
}

----

== What's next

There's a proposed Google Summer of Code project: link:https://jenkins.io/projects/gsoc/2020/project-ideas/github-checks/[GitHub Checks API for Jenkins Plugins]. 

This is looking at integrating with the link:https://developer.github.com/v3/checks/[checks api], with a focus on reporting issues found using the link:https://plugins.jenkins.io/warnings-ng/[warnings-ng plugin] directly onto the GitHub pull request, along with test results summary on GitHub.

If you want to get involved with this, join the link:https://gitter.im/jenkinsci/gsoc-sig[gitter channel] and ask how you can help.

> There are already students sending in their proposals.

Author: ```suggestion
```
